### ISSUE TITLE
Scope of project?

### ISSUE THREAD

**Issue by sebastiansam55** [dp0s/evdev_prepared#2]
I use `evdev` in a number of different projects and have been procrastinating making a helper library. Can I ask what you think that the scope of the project will end up being?

A few methods that I would like to see implemented (have not gone through the 0.2.0 release):
- Test that user is in the `input` group (or whatever metric measures that the user has full access to use the library)
- Translation from strings to evdev keycodes (including most keys)
  - ie: `to_evdev("Str<Home>")` would output a list like
    - [[shift, s_code], t_code, r_code, home_code]
      - I have semi implemented this in some of my projects (see end)
- Better device grabbing;
  - Be able to grab devices by their string names;
    -  I have semi implemented this in some of my projects (see end)
- Some way to identify devices by type;
  - Either attempt to find mouse/keyboard/controller in the plaintext name of the device or
  - Try to guess device type via looking at the capabilites of the item.
- Better capabilities
  - Prebuilt capability objects;
    - all events
    - keyboard preset
    - mouse preset
    - controller preset
    - etc. 

I have a partially complete key map that you can find attached, hopefully it is useful. [keymap.txt](https://github.com/dp0s/evdev_prepared/files/6206856/keymap.txt)

The device grabbing method that I currently use supports a string that can either be just the `event00` number, the `/dev/input/event` path or the complete full text name of the object (no partial matching).

https://github.com/sebastiansam55/uinput-macropad/blob/main/macropad.py#L11-L42

**Comment by dp0s** [comment]
Hey Sebastiansam, 

The 0.2.0 release did not add new features, it mainly updated the evdev version. 

I will answer your points step by step.

- translation from strings to ecodes (and vice versa) is actually handled inside my Dpowers project (to which evdev_prepared is a backend).
It is not documented yet, but you can try:

```
from Dpowers import keyb,mouse,hook, sleep
keyb.adapt("evdev")
keyb.tap("space")
mouse.adapt("evdev")
mouse.tap("left")
hook.adapt("evdev")
with hook.keys(print): sleep(5)
```

**Comment by dp0s** [comment]
- grab devices by their names or by their category is already implemented in evdev_prepared. See evdev_prepared.device_control.DeviceSelector class

The categories are defined in the module evdev_prepared.input_dev 

All of this is used by the Dpowers project which supports selecting custom devices. (source code in Dpowers.events.hookpower.adapt_evdev)

**Comment by sebastiansam55** [comment]
Nice I'll check dpowers out. 

My interest in this project mostly stems from my work on the [autokey](https://github.com/autokey/autokey/) project (which you should also checkout if you have not heard of it, we are always looking for help!) and wanting to bring `uinput` support to the project through evdev.

That starts with having a reliable way to translate from keycodes to strings and back.

Would you be open to a PR implementing that kind of functionality?

**Comment by dp0s** [comment]
Good to know you are working on autokey. I used it in 2017 for about 6 months before starting the work on Dpowers which replaced autokey for me.

As far as I can remember, autokey has 3 components (at least in 2017):
1 gui
2 capability to define hotkeys and hotstrings to trigger python scripts
3 an own interface (e.g. for sending keystrokes), usable from within autokey. 

My thoughts about it:
1 the gui looks really nice, but I prefer to not use one. Dpowers uses function decorators to define hotkeys instead. 
2 I'm actually ja bit jealous of autokey 's hotkey and hotstring capabilities, because that's the one point where it is superior to Dpowers. I actually tried to copy some code, but it was so complicated I couldn't figure it out. 
3 as far as I know, Dpowers support all features that the autokey interface has. Need to check for updates though. 

The good thing is that the Dpowers interface should be usable from within autokey scripts right away. That's what I like about open source :-) 

I would also like to use some autokey libraries as additional backends for Dpowers, but as far as I know, those libraries cannot be used outside of autokey?

**Comment by dp0s** [comment]
About your question of translating keycodes:

Creating a dictionary between ecodes and key names can be done like the following:
```
from evdev.ecodes import KEY, EV_KEY

names = {}

def iterator(dic):
    for a, b in dic.items():
        if isinstance(b, (tuple, list)):
            for c in b: yield a,c
        else:
            yield a, b

for integer, name in iterator(KEY):
    n = name[4:]
    if len(n)==1 and n.isalpha(): n=n.lower()
    names[n] = integer
```

(taken from Dpowers.events.sending.keybpower.adapt_evdev)

If you want to use evdev_prepared without Dpowers, I could try to migrate that from Dpowers to evdev_prepared.

EDIT: This doesn't solve the complete translation problem  -- see next comment.

**Comment by dp0s** [comment]
I just had a look at your macropad project -- it is similar in functionality compared to what I'm trying to offer in keybpower subpackage. Lets see what we can learn from each other :-)

I also noticed that the Dpowers are not yet able to send capital letters using evdev. (keyb.tap("A") raises an exception while keyb.tap("a") works fine). For my personal use I employ pynput as backend (via keyb.adapt("pynput")), which is able to send capital letters via Xlib.  But of course, it would be nice to also do that via evdev. 

For this purpose, you can have a look at Dhelpers.KeyboardLayout  subpackage (also part of the Dpowers library). There is a Layout class and a collection of json-Files containing keyboard layout tables for common languages.  The keyboard Layout class is able to translate between languages (necessary for evdev because I type on a German keyboard, but evdev still prints the english names for each key.)  It shouldn't be too difficult to extend it's functionality to also be able to send shifted keys via evdev for any keyboard Layout.  That would also be a useful enhancement for the Dpowers.

Btw, I know that the author of pynput is also working on an evdev backend -- Have you checked if he has solved the problem already?

Edit: there seems to be a solution already here:
https://github.com/moses-palmer/pynput/tree/master/lib/pynput/keyboard/_uinput.py

Maybe I will still implement another shifting/translation algorithm inside Dpowers at some point, allowing the user to manually adapt it.

**Comment by dp0s** [comment]
> Test that user is in the input group (or whatever metric measures that the user has full access to use the library)

That's something I definitely want to implement in evdev_prepared. When I installed evdev for the first time, it was really tedious to make it work without root permissions.

Probably we should open another issue to discuss that.

**Comment by sebastiansam55** [comment]
Yeah keyboards/keycodes/scancodes are super complicated if you want to have universal support. I have put some final touches on the code that I will be using in one of my projects; you can see the tests here:
https://github.com/sebastiansam55/python-keycodes/blob/main/tests/test_keycodes.py

The idea is to return a list that the code can then iterate over to send events;
```python
keycodes.list_to_evdev(["LeftControl", "R"])  #returns [29, [[42], 19]]
keys = keycodes.str_to_evdev("this is a test") #[20, 35, 23, 31, 57, 23, 31, 57, 30, 57, 20, 18, 31, 20]
hotkey = keycodes.hotkey_to_evdev([["LeftControl", "LeftShift"], "a"]) # [[29, 42], 30]
# beyond here is more pseudo code than anything else.
for key in keys:
  if key is list:
    for k in key[0]: # press the modifiers down
      uinput.write(k, down)
    for k in key[1:]:
      uinput.write(k) #down, up
    for k in key[0]:
      uinput.write(k, up)
  else:
    uinput.write(key)

```

**Comment by dp0s** [comment]
It seems translating the keycodes can be done in many ways, depending on the needs of the current project you're working on.

For me it would be useful to know:
- Do you want to use evdev_prepared as a backend dependency to your project? Alternatively, you can of course modify the code and put it into your own project as it suits you (if your project is under GPLv3).
- Is it an option for you to use the already existing functions in Dpowers?  It has almost all the features you describe and I could easily implement the remaining (such as converting A to shift+a -- see my post from further above). But maybe you prefer to use your own code, so you have full control over the behavior.

I have added documentation to Dpowers.keyb:
https://dpowers.readthedocs.io/en/latest/reference/power.html#keyb-keyboardadaptor


Your examples using Dpowers syntax:
```
from Dpowers import keyb
keyb.adapt("evdev")
keyb.comb("CtrlL","r")
keyb.text("this is a test")
keyb.comb("CtrlL","ShiftL", "a")
```

**Comment by dp0s** [comment]
@sebastiansam55:  I have implemented a key shifting mechanism in the Dpowers supporting evdev. I don't know, if that is what you need, but you can create dictionaries for evdev as follows:

```
from Dpowers import keyb
keyb.adapt("evdev")

# default US layout:
print(keyb._effective_dict_expanded())


# German keyboard layout:
keyb.set_layout("de")
print(keyb._effective_dict_expanded())
```


The resulting evdev translation table for US layout (including "shift" as modifier with evdev code 43):

```
{'0': [11], '1': [2], '<': [42, 51], '2': [3], '3': [4], '4': [5], '5': [6], '6': [7], '7': [8], '8': [9], '9': [10], 'a': [30], "'": [40], 'appselect': [580], 'b': [48], '\\': [43], 'Backspace': [14], 'c': [46], 'calc': [140], 'Caps_Lock': [58], ',': [51], 'compose': [127], 'd': [32], 'Delete': [111], '$': [42, 5], '.': [52], 'down': [108], 'e': [18], 'End': [353], 'Enter': [28], '=': [13], 'Esc': [1], 'f': [33], 'F1': [59], 'F10': [68], 'F11': [87], 'F12': [88], 'F2': [60], 'F3': [61], 'F4': [62], 'F5': [63], 'F6': [64], 'F7': [65], 'F8': [66], 'F9': [67], 'Home': [102], 'g': [34], '`': [41], 'h': [35], 'i': [23], 'Insert': [110], 'j': [36], 'k': [37], 'p_0': [82], 'p_1': [79], 'p_2': [80], 'p_3': [81], 'p_4': [75], 'p_5': [76], 'p_6': [77], 'p_7': [71], 'p_8': [72], 'p_9': [73], 'p_multiply': [55], 'p_dot': [83], 'p_enter': [96], 'p_subtract': [74], 'p_add': [78], 'p_divide': [98], 'l': [38], 'left': [105], 'Alt': [56], '{': [42, 26], 'CtrlL': [29], 'Win': [125], 'ShiftL': [42], 'Linefeed': [101], 'm': [50], 'Menu': [139], '-': [12], 'mute': [113], 'n': [49], 'PageDown': [109], 'nextsong': [163], 'Num_Lock': [69], 'o': [24], 'p': [25], 'PageUp': [104], 'Pause': [119], 'play': [164], 'previoussong': [165], 'Print': [99], 'q': [16], '?': [42, 53], 'r': [19], 'right': [106], 'AltGr': [100], '}': [42, 27], 'CtrlR': [97], 'ShiftR': [54], 's': [31], 'Scroll_Lock': [70], ';': [39], '/': [53], 'sleep': [142], ' ': [57], 'stop': [166], 't': [20], 'Tab': [15], 'touchpad_off': [532], 'touchpad_on': [531], 'u': [22], 'up': [103], 'v': [47], 'volumedown': [114], 'volumeup': [115], 'w': [17], 'wlan': [238], 'x': [45], 'y': [21], '¥': [124], 'z': [44], '~': [42, 41], '!': [42, 2], '@': [42, 3], '#': [42, 4], '%': [42, 6], '^': [42, 7], '&': [42, 8], '*': [42, 9], '(': [42, 10], ')': [42, 11], '_': [42, 12], '+': [42, 13], 'Q': [42, 16], 'W': [42, 17], 'E': [42, 18], 'R': [42, 19], 'T': [42, 20], 'Y': [42, 21], 'U': [42, 22], 'I': [42, 23], 'O': [42, 24], 'P': [42, 25], '[': [26], ']': [27], 'A': [42, 30], 'S': [42, 31], 'D': [42, 32], 'F': [42, 33], 'G': [42, 34], 'H': [42, 35], 'J': [42, 36], 'K': [42, 37], 'L': [42, 38], ':': [42, 39], '"': [42, 40], 'Z': [42, 44], 'X': [42, 45], 'C': [42, 46], 'V': [42, 47], 'B': [42, 48], 'N': [42, 49], 'M': [42, 50], '>': [42, 52], '|': [42, 43], '10channelsdown': [441], '10channelsup': [440], '3d_mode': [623], 'ab': [406], 'addressbook': [429], 'again': [129], 'als_toggle': [560], 'alterase': [222], 'angle': [371], 'archive': [361], 'assistant': [583], 'attendant_off': [540], 'attendant_on': [539], 'attendant_toggle': [541], 'audio': [392], 'audio_desc': [622], 'aux': [390], 'back': [158], 'bassboost': [209], 'battery': [236], 'blue': [401], 'bluetooth': [237], 'bookmarks': [156], 'break': [411], 'brightnessdown': [224], 'brightnessup': [225], 'brightness_auto': [244], 'brightness_zero': [244], 'brightness_cycle': [243], 'brightness_max': [593], 'brightness_min': [592], 'brightness_toggle': [431], 'displaytoggle': [431], 'brl_dot1': [497], 'brl_dot10': [506], 'brl_dot2': [498], 'brl_dot3': [499], 'brl_dot4': [500], 'brl_dot5': [501], 'brl_dot6': [502], 'brl_dot7': [503], 'brl_dot8': [504], 'brl_dot9': [505], 'buttonconfig': [576], 'calendar': [397], 'camera': [212], 'camera_down': [536], 'camera_focus': [528], 'camera_left': [537], 'camera_right': [538], 'camera_up': [535], 'camera_zoomin': [533], 'camera_zoomout': [534], 'cancel': [223], 'cd': [383], 'channel': [363], 'channeldown': [403], 'channelup': [402], 'chat': [216], 'clear': [355], 'close': [206], 'closecd': [160], 'cnt': [768], 'coffee': [152], 'screenlock': [152], 'computer': [157], 'config': [171], 'connect': [218], 'context_menu': [438], 'controlpanel': [579], 'copy': [133], 'cut': [137], 'cyclewindows': [154], 'dashboard': [204], 'data': [631], 'database': [426], 'deletefile': [146], 'del_eol': [448], 'del_eos': [449], 'del_line': [451], 'digits': [413], 'direction': [153], 'rotate_display': [153], 'directory': [394], 'display_off': [245], 'documents': [235], 'dvd': [389], 'edit': [176], 'editor': [422], 'ejectcd': [161], 'ejectclosecd': [162], 'email': [215], 'epg': [365], 'euro': [435], 'exit': [174], 'f13': [183], 'f14': [184], 'f15': [185], 'f16': [186], 'f17': [187], 'f18': [188], 'f19': [189], 'f20': [190], 'f21': [191], 'f22': [192], 'f23': [193], 'f24': [194], 'fastforward': [208], 'fastreverse': [629], 'favorites': [364], 'file': [144], 'finance': [219], 'first': [404], 'fn': [464], 'fn_1': [478], 'fn_2': [479], 'fn_b': [484], 'fn_d': [480], 'fn_e': [481], 'fn_esc': [465], 'fn_f': [482], 'fn_f1': [466], 'fn_f10': [475], 'fn_f11': [476], 'fn_f12': [477], 'fn_f2': [467], 'fn_f3': [468], 'fn_f4': [469], 'fn_f5': [470], 'fn_f6': [471], 'fn_f7': [472], 'fn_f8': [473], 'fn_f9': [474], 'fn_s': [483], 'forward': [159], 'forwardmail': [233], 'frameback': [436], 'frameforward': [437], 'front': [132], 'games': [417], 'goto': [354], 'graphicseditor': [424], 'green': [399], 'hangeul': [122], 'hanguel': [122], 'hanja': [123], 'help': [138], 'henkan': [92], 'hiragana': [91], 'homepage': [172], 'hp': [211], 'images': [442], 'info': [358], 'ins_line': [450], 'iso': [170], 'journal': [578], 'katakana': [90], 'katakanahiragana': [93], 'kbdillumdown': [229], 'kbdillumtoggle': [228], 'kbdillumup': [230], 'kbdinputassist_accept': [612], 'kbdinputassist_cancel': [613], 'kbdinputassist_next': [609], 'kbdinputassist_nextgroup': [611], 'kbdinputassist_prev': [608], 'kbdinputassist_prevgroup': [610], 'keyboard': [374], 'kpcomma': [121], 'kpequal': [117], 'kpjpcomma': [95], 'kpleftparen': [179], 'kpplusminus': [118], 'kprightparen': [180], 'language': [368], 'last': [405], 'left_down': [617], 'left_up': [616], 'lights_toggle': [542], 'list': [395], 'logoff': [433], 'macro': [112], 'mail': [155], 'max': [767], 'media': [226], 'media_repeat': [439], 'media_top_menu': [619], 'memo': [396], 'messenger': [430], 'mhp': [367], 'micmute': [248], 'mode': [373], 'move': [175], 'mp3': [391], 'msdos': [151], 'muhenkan': [94], 'new': [181], 'news': [427], 'next_favorite': [624], 'numeric_0': [512], 'numeric_1': [513], 'numeric_11': [620], 'numeric_12': [621], 'numeric_2': [514], 'numeric_3': [515], 'numeric_4': [516], 'numeric_5': [517], 'numeric_6': [518], 'numeric_7': [519], 'numeric_8': [520], 'numeric_9': [521], 'numeric_a': [524], 'numeric_b': [525], 'numeric_c': [526], 'numeric_d': [527], 'numeric_pound': [523], 'numeric_star': [522], 'ok': [352], 'onscreen_keyboard': [632], 'open': [134], 'option': [357], 'paste': [135], 'pausecd': [201], 'pause_record': [626], 'pc': [376], 'phone': [169], 'power': [116], 'power2': [356], 'presentation': [425], 'previous': [412], 'prog1': [148], 'prog2': [149], 'prog3': [202], 'prog4': [203], 'program': [362], 'props': [130], 'pvr': [366], 'radio': [385], 'record': [167], 'red': [398], 'redo': [182], 'refresh': [173], 'reply': [232], 'reserved': [0], 'restart': [408], 'rewind': [168], 'rfkill': [247], 'rightmeta': [126], 'right_down': [615], 'right_up': [614], 'ro': [89], 'root_menu': [618], 'sat': [381], 'sat2': [382], 'save': [234], 'scale': [120], 'screen': [375], 'screensaver': [581], 'scrolldown': [178], 'scrollup': [177], 'search': [217], 'send': [231], 'sendfile': [145], 'setup': [141], 'shop': [221], 'shuffle': [410], 'slow': [409], 'slowreverse': [630], 'sound': [213], 'spellcheck': [432], 'sport': [220], 'spreadsheet': [423], 'stop_record': [625], 'subtitle': [370], 'suspend': [205], 'switchvideomode': [227], 'tape': [384], 'taskmanager': [577], 'teen': [414], 'text': [388], 'time': [359], 'title': [369], 'touchpad_toggle': [530], 'tuner': [386], 'tv': [377], 'tv2': [378], 'twen': [415], 'undo': [131], 'unknown': [240], 'unmute': [628], 'uwb': [239], 'vcr': [379], 'vcr2': [380], 'vendor': [360], 'video': [393], 'videophone': [416], 'video_next': [241], 'video_prev': [242], 'vod': [627], 'voicecommand': [582], 'voicemail': [428], 'wakeup': [143], 'wimax': [246], 'wwan': [246], 'wordprocessor': [421], 'wps_button': [529], 'www': [150], 'xfer': [147], 'yellow': [400], 'zenkakuhankaku': [85], 'zoom': [372], 'zoomin': [418], 'zoomout': [419], 'zoomreset': [420]}
```


And for German layout (including "shift" as modifier as well as "AltGr" with evdev code 100):

```
{'0': [11], '1': [2], '<': [170], '2': [3], '3': [4], '4': [5], '5': [6], '6': [7], '7': [8], '8': [9], '9': [10], 'a': [30], "'": [42, 43], 'appselect': [580], 'b': [48], '\\': [100, 12], 'Backspace': [14], 'c': [46], 'calc': [140], 'Caps_Lock': [58], ',': [51], 'compose': [127], 'd': [32], 'Delete': [111], '$': [42, 5], '.': [52], 'down': [108], 'e': [18], 'End': [353], 'Enter': [28], '=': [42, 11], 'Esc': [1], 'f': [33], 'F1': [59], 'F10': [68], 'F11': [87], 'F12': [88], 'F2': [60], 'F3': [61], 'F4': [62], 'F5': [63], 'F6': [64], 'F7': [65], 'F8': [66], 'F9': [67], 'Home': [102], 'g': [34], '`': [42, 13], 'h': [35], 'i': [23], 'Insert': [110], 'j': [36], 'k': [37], 'p_0': [82], 'p_1': [79], 'p_2': [80], 'p_3': [81], 'p_4': [75], 'p_5': [76], 'p_6': [77], 'p_7': [71], 'p_8': [72], 'p_9': [73], 'p_multiply': [55], 'p_dot': [83], 'p_enter': [96], 'p_subtract': [74], 'p_add': [78], 'p_divide': [98], 'l': [38], 'left': [105], 'Alt': [56], '{': [100, 8], 'CtrlL': [29], 'Win': [125], 'ShiftL': [42], 'Linefeed': [101], 'm': [50], 'Menu': [139], '-': [53], 'mute': [113], 'n': [49], 'PageDown': [109], 'nextsong': [163], 'Num_Lock': [69], 'o': [24], 'p': [25], 'PageUp': [104], 'Pause': [119], 'play': [164], 'previoussong': [165], 'Print': [99], 'q': [16], '?': [42, 12], 'r': [19], 'right': [106], 'AltGr': [100], '}': [100, 11], 'CtrlR': [97], 'ShiftR': [54], 's': [31], 'Scroll_Lock': [70], ';': [42, 51], '/': [42, 8], 'sleep': [142], ' ': [57], 'stop': [166], 't': [20], 'Tab': [15], 'touchpad_off': [532], 'touchpad_on': [531], 'u': [22], 'up': [103], 'v': [47], 'volumedown': [114], 'volumeup': [115], 'w': [17], 'wlan': [238], 'x': [45], 'y': [44], '¥': [42, 100, 21], 'z': [21], '!': [42, 2], '¹': [100, 2], '¡': [42, 100, 2], '"': [42, 3], '²': [100, 3], '§': [42, 4], '³': [100, 4], '£': [42, 100, 4], '¼': [100, 5], '¤': [42, 100, 5], '%': [42, 6], '½': [100, 6], '&': [42, 100, 37], '¬': [100, 7], '(': [42, 9], '[': [100, 9], ')': [42, 10], ']': [100, 10], '±': [42, 100, 10], '°': [42, 41], 'ß': [42, 100, 31], '¿': [42, 100, 12], '´': [13], '¸': [100, 13], 'Q': [42, 16], '@': [100, 16], 'W': [42, 17], 'E': [42, 18], 'R': [42, 19], '¶': [100, 19], '®': [42, 100, 19], 'T': [42, 20], 'Z': [42, 21], 'U': [42, 22], 'I': [42, 23], 'O': [42, 24], 'ø': [100, 24], 'P': [42, 25], 'þ': [100, 25], 'ü': [26], '¨': [100, 26], '+': [27], '*': [42, 27], '~': [100, 27], '¯': [42, 100, 27], 'A': [42, 30], 'æ': [100, 30], 'S': [42, 31], 'D': [42, 32], 'ð': [100, 32], 'F': [42, 33], 'ª': [42, 100, 33], 'G': [42, 34], 'H': [42, 35], 'J': [42, 36], 'K': [42, 37], 'L': [42, 38], 'ö': [39], 'ä': [40], '^': [41], '#': [43], 'Y': [42, 44], '»': [100, 44], 'X': [42, 45], '«': [100, 45], 'C': [42, 46], '¢': [100, 46], '©': [42, 100, 46], 'V': [42, 47], 'B': [42, 48], 'N': [42, 49], 'M': [42, 50], 'µ': [100, 50], 'º': [42, 100, 50], '·': [100, 51], '×': [42, 100, 51], ':': [42, 52], '÷': [42, 100, 52], '_': [42, 53], '>': [42, 170], '|': [100, 170], 'p_delete': [83], '10channelsdown': [441], '10channelsup': [440], '3d_mode': [623], 'ab': [406], 'addressbook': [429], 'again': [129], 'als_toggle': [560], 'alterase': [222], 'angle': [371], 'archive': [361], 'assistant': [583], 'attendant_off': [540], 'attendant_on': [539], 'attendant_toggle': [541], 'audio': [392], 'audio_desc': [622], 'aux': [390], 'back': [158], 'bassboost': [209], 'battery': [236], 'blue': [401], 'bluetooth': [237], 'bookmarks': [156], 'break': [411], 'brightnessdown': [224], 'brightnessup': [225], 'brightness_auto': [244], 'brightness_zero': [244], 'brightness_cycle': [243], 'brightness_max': [593], 'brightness_min': [592], 'brightness_toggle': [431], 'displaytoggle': [431], 'brl_dot1': [497], 'brl_dot10': [506], 'brl_dot2': [498], 'brl_dot3': [499], 'brl_dot4': [500], 'brl_dot5': [501], 'brl_dot6': [502], 'brl_dot7': [503], 'brl_dot8': [504], 'brl_dot9': [505], 'buttonconfig': [576], 'calendar': [397], 'camera': [212], 'camera_down': [536], 'camera_focus': [528], 'camera_left': [537], 'camera_right': [538], 'camera_up': [535], 'camera_zoomin': [533], 'camera_zoomout': [534], 'cancel': [223], 'cd': [383], 'channel': [363], 'channeldown': [403], 'channelup': [402], 'chat': [216], 'clear': [355], 'close': [206], 'closecd': [160], 'cnt': [768], 'coffee': [152], 'screenlock': [152], 'computer': [157], 'config': [171], 'connect': [218], 'context_menu': [438], 'controlpanel': [579], 'copy': [133], 'cut': [137], 'cyclewindows': [154], 'dashboard': [204], 'data': [631], 'database': [426], 'deletefile': [146], 'del_eol': [448], 'del_eos': [449], 'del_line': [451], 'digits': [413], 'direction': [153], 'rotate_display': [153], 'directory': [394], 'display_off': [245], 'documents': [235], 'dvd': [389], 'edit': [176], 'editor': [422], 'ejectcd': [161], 'ejectclosecd': [162], 'email': [215], 'epg': [365], 'euro': [435], 'exit': [174], 'f13': [183], 'f14': [184], 'f15': [185], 'f16': [186], 'f17': [187], 'f18': [188], 'f19': [189], 'f20': [190], 'f21': [191], 'f22': [192], 'f23': [193], 'f24': [194], 'fastforward': [208], 'fastreverse': [629], 'favorites': [364], 'file': [144], 'finance': [219], 'first': [404], 'fn': [464], 'fn_1': [478], 'fn_2': [479], 'fn_b': [484], 'fn_d': [480], 'fn_e': [481], 'fn_esc': [465], 'fn_f': [482], 'fn_f1': [466], 'fn_f10': [475], 'fn_f11': [476], 'fn_f12': [477], 'fn_f2': [467], 'fn_f3': [468], 'fn_f4': [469], 'fn_f5': [470], 'fn_f6': [471], 'fn_f7': [472], 'fn_f8': [473], 'fn_f9': [474], 'fn_s': [483], 'forward': [159], 'forwardmail': [233], 'frameback': [436], 'frameforward': [437], 'front': [132], 'games': [417], 'goto': [354], 'graphicseditor': [424], 'green': [399], 'hangeul': [122], 'hanguel': [122], 'hanja': [123], 'help': [138], 'henkan': [92], 'hiragana': [91], 'homepage': [172], 'hp': [211], 'images': [442], 'info': [358], 'ins_line': [450], 'iso': [170], 'journal': [578], 'katakana': [90], 'katakanahiragana': [93], 'kbdillumdown': [229], 'kbdillumtoggle': [228], 'kbdillumup': [230], 'kbdinputassist_accept': [612], 'kbdinputassist_cancel': [613], 'kbdinputassist_next': [609], 'kbdinputassist_nextgroup': [611], 'kbdinputassist_prev': [608], 'kbdinputassist_prevgroup': [610], 'keyboard': [374], 'kpcomma': [121], 'kpequal': [117], 'kpjpcomma': [95], 'kpleftparen': [179], 'kpplusminus': [118], 'kprightparen': [180], 'language': [368], 'last': [405], 'left_down': [617], 'left_up': [616], 'lights_toggle': [542], 'list': [395], 'logoff': [433], 'macro': [112], 'mail': [155], 'max': [767], 'media': [226], 'media_repeat': [439], 'media_top_menu': [619], 'memo': [396], 'messenger': [430], 'mhp': [367], 'micmute': [248], 'mode': [373], 'move': [175], 'mp3': [391], 'msdos': [151], 'muhenkan': [94], 'new': [181], 'news': [427], 'next_favorite': [624], 'numeric_0': [512], 'numeric_1': [513], 'numeric_11': [620], 'numeric_12': [621], 'numeric_2': [514], 'numeric_3': [515], 'numeric_4': [516], 'numeric_5': [517], 'numeric_6': [518], 'numeric_7': [519], 'numeric_8': [520], 'numeric_9': [521], 'numeric_a': [524], 'numeric_b': [525], 'numeric_c': [526], 'numeric_d': [527], 'numeric_pound': [523], 'numeric_star': [522], 'ok': [352], 'onscreen_keyboard': [632], 'open': [134], 'option': [357], 'paste': [135], 'pausecd': [201], 'pause_record': [626], 'pc': [376], 'phone': [169], 'power': [116], 'power2': [356], 'presentation': [425], 'previous': [412], 'prog1': [148], 'prog2': [149], 'prog3': [202], 'prog4': [203], 'program': [362], 'props': [130], 'pvr': [366], 'radio': [385], 'record': [167], 'red': [398], 'redo': [182], 'refresh': [173], 'reply': [232], 'reserved': [0], 'restart': [408], 'rewind': [168], 'rfkill': [247], 'rightmeta': [126], 'right_down': [615], 'right_up': [614], 'ro': [89], 'root_menu': [618], 'sat': [381], 'sat2': [382], 'save': [234], 'scale': [120], 'screen': [375], 'screensaver': [581], 'scrolldown': [178], 'scrollup': [177], 'search': [217], 'send': [231], 'sendfile': [145], 'setup': [141], 'shop': [221], 'shuffle': [410], 'slow': [409], 'slowreverse': [630], 'sound': [213], 'spellcheck': [432], 'sport': [220], 'spreadsheet': [423], 'stop_record': [625], 'subtitle': [370], 'suspend': [205], 'switchvideomode': [227], 'tape': [384], 'taskmanager': [577], 'teen': [414], 'text': [388], 'time': [359], 'title': [369], 'touchpad_toggle': [530], 'tuner': [386], 'tv': [377], 'tv2': [378], 'twen': [415], 'undo': [131], 'unknown': [240], 'unmute': [628], 'uwb': [239], 'vcr': [379], 'vcr2': [380], 'vendor': [360], 'video': [393], 'videophone': [416], 'video_next': [241], 'video_prev': [242], 'vod': [627], 'voicecommand': [582], 'voicemail': [428], 'wakeup': [143], 'wimax': [246], 'wwan': [246], 'wordprocessor': [421], 'wps_button': [529], 'www': [150], 'xfer': [147], 'yellow': [400], 'zenkakuhankaku': [85], 'zoom': [372], 'zoomin': [418], 'zoomout': [419], 'zoomreset': [420], '⅛': [42, 100, 3], '⅜': [42, 100, 6], '⅝': [42, 100, 7], '⅞': [42, 100, 8], '™': [42, 100, 9], 'ogonek': [42, 100, 13], 'ω': [42, 100, 16], 'ł': [42, 100, 38], '€': [42, 100, 18], 'ŧ': [42, 100, 20], '←': [100, 21], '↓': [100, 22], '↑': [42, 100, 22], '→': [100, 23], 'ı': [42, 100, 23], 'Ø': [42, 100, 24], 'Þ': [42, 100, 25], 'Ü': [42, 26], 'abovering': [42, 100, 26], 'Æ': [42, 100, 30], 'ſ': [100, 31], 'Ð': [42, 100, 32], 'đ': [100, 33], 'ŋ': [42, 100, 34], 'ħ': [42, 100, 35], 'belowdot': [42, 100, 39], 'abovedot': [42, 100, 36], 'ĸ': [100, 37], 'Ö': [42, 39], 'doubleacute': [100, 39], 'Ä': [42, 40], 'caron': [42, 100, 40], '′': [100, 41], '″': [42, 100, 41], '’': [42, 100, 49], 'breve': [42, 100, 43], '›': [42, 100, 44], '‹': [42, 100, 45], '„': [100, 47], '‚': [42, 100, 47], '“': [100, 48], '‘': [42, 100, 48], '”': [100, 49], '…': [100, 52], '–': [100, 53], '—': [42, 100, 53], 'belowmacron': [42, 100, 170], 'kp_comma': [42, 83]}
```
